core/src/main/java/com/linecorp/armeria/client/retry/RetryStrategyBuilder.java

```diff
@@ -0,0 +1,121 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Predicate;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.HttpStatusClass;
+import com.linecorp.armeria.common.logging.RequestLogProperty;
+import com.linecorp.armeria.common.util.Exceptions;
+
+/**
+ * A builder which creates a {@link RetryStrategy}.
+ */
+public final class RetryStrategyBuilder {
+
+    private static final CompletableFuture<Backoff> NULL_BACKOFF = CompletableFuture.completedFuture(null);
+    private static final RetryStrategy[] EMPTY_RETRY_STRATEGIES = new RetryStrategy[0];
+
+    private final ImmutableList.Builder<RetryStrategy> retryStrategiesBuilder = ImmutableList.builder();
+
+    RetryStrategyBuilder() {}
+
+    /**
+     * Adds a {@link RetryStrategy}.
+     */
+    public RetryStrategyBuilder add(RetryStrategy retryStrategy) {
+        retryStrategiesBuilder.add(requireNonNull(retryStrategy, "retryStrategy"));
+        return this;
+    }
+
+    /**
+     * Adds a {@link RetryRule}.
+     *
+     * <p><pre>{@code
+     * RetryStrategy.builder()
+     *              .rule(RetryRule.onStatus(HttpStatus.SERVICE_UNAVAILABLE)
+     *                             .onException(ex -> ex instanceof ClosedSessionException)
+     *                             .onMethod(HttpMethod.GET)
+     *                             .thenBackOff(myBackoff))
+     *              .build();
+     * }</pre>
+     */
+    public RetryStrategyBuilder rule(RetryRule retryRule) {
+        return add(build(requireNonNull(retryRule, "retryRule")));
+    }
+
+    /**
+     * Returns a newly-created {@link RetryStrategy} based on the strategies set so far.
+     */
+    public RetryStrategy build() {
+        final RetryStrategy[] retryStrategies = retryStrategiesBuilder.build().toArray(EMPTY_RETRY_STRATEGIES);
+        checkState(retryStrategies.length > 0, "at least one retry strategy should be set");
+
+        if (retryStrategies.length == 1) {
+            return retryStrategies[0];
+        }
+
+        return (ctx, cause) -> {
+            for (RetryStrategy retryStrategy : retryStrategies) {
+                final CompletionStage<Backoff> backoff = retryStrategy.shouldRetry(ctx, cause);
```

Comment:
I think this will not work because:
- the customized strategy cannot return `NULL_BACKOFF`
- the second strategy might return valid backoff meanwhile the first strategy completes the future with `null`
How about using `allOf`?
```java
final List<RetryStrategy> retryStrategies = retryStrategiesBuilder.build();
checkState(!retryStrategies.isEmpty(), "at least one retry strategy should be set");

if (retryStrategies.size() == 1) {
    return retryStrategies.get(0);
}

return (ctx, cause) -> {
    final List<CompletionStage<Backoff>> futures = new ArrayList<>(retryStrategies.size());
    final CompletableFuture<Void> allOfFuture = CompletableFuture.allOf(
            retryStrategies.stream()
                           .map(retryStrategy -> {
                               final CompletionStage<Backoff> future =
                                       retryStrategy.shouldRetry(ctx, cause);
                               futures.add(future);
                               return future;
                           })
                           .toArray(CompletableFuture[]::new));
    return allOfFuture.handle((unused, unused1) -> {
        for (CompletionStage<Backoff> future : futures) {
            // Perhaps is it time to change the return type of shouldRetry to CompletableFuture?
            final Backoff backoff = future.toCompletableFuture().join();
            if (backoff != null) {
                return backoff;
            }
        }
        return null;
    });
};
```

Reply:
Or perhaps we can implement `anyOfNonNull(futures...)`.

Reply:
> the customized strategy cannot return NULL_BACKOFF

Oops... thank!

> How about using allOf?

I don't think we can use `allOf` here. Since it invokes all `shouldRetry`. If a user adds some side effect such as retry metric, `allOf` cause a problem.

Let me get backoffs lazily.